# 4.6. Exercise Example

# 
This script contains the following points:# 
1. Import the data sets into Jupyte# r
2. Check the dimensions of the imported datafram# es
3. Merge the datafra# mes
4. Use the indicator argument to check whether there was a full match between the two datafr# ames
5. Export data in pickle ormat¶ormat¶

# Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import the data sets into Jupyter

In [19]:
path = r'C:\Users\João Pedro\MAR-Instacart Basket Analysis'

In [20]:
path

'C:\\Users\\João Pedro\\MAR-Instacart Basket Analysis'

# Import dataset orders_products_prior.csv

In [21]:
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders_products_prior.csv'), index_col = False)

# Import dataset orders_checked.csv

In [22]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index_col = False)

# Check dimension of the imported dataframes

In [23]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [24]:
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id eval_set  order_number  \
0             0           0   2539329        1    prior             1   
1             1           1   2398795        1    prior             2   
2             2           2    473747        1    prior             3   
3             3           3   2254736        1    prior             4   
4             4           4    431534        1    prior             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  first_order  
0                   2                  8                     NaN         True  
1                   3                  7                    15.0        False  
2                   3                 12                    21.0        False  
3                   4                  7                    29.0        False  
4                   4                 15                    28.0        False

In [25]:
df_ords.shape

(3421083, 10)

# Delete unnencessary columns

In [26]:
df_ords = df_ords.drop(['Unnamed: 0.1', 'Unnamed: 0', 'eval_set'], axis=1)

# Check output

In [27]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  
0                     NaN         True  
1                    15.0        False  
2                    21.0        False  
3                    29.0        False  
4                    28.0        False

In [28]:
df_ords.shape

(3421083, 7)

In [29]:
df_ords_prior.shape

(32434489, 4)

# Merge the dataframes

In [30]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [31]:
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

# Use the indicator argument to check wether there was a full mathc in the 2 dataframes

In [32]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

## After the merge, the resulting dataframe contains 32,434,489 rows, each incorporating information from both input data sets due to our use of an inner join for the project's objectives. ##

# Export data in pickle format

In [33]:
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))

# Export modified df_ords dataframe

In [34]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_final.csv'))